turbopack/crates/turbopack-core/src/module_graph/merged_modules.rs

```diff
@@ -0,0 +1,584 @@
+        let module_graph = module_graph.await?;
+
+        let graphs = module_graph.graphs.iter().try_join().await?;
+        let module_count = graphs.iter().map(|g| g.graph.node_count()).sum::<usize>();
+        span.record("module_count", module_count);
+
+        // Use all entries from all graphs
+        let entries = graphs
+            .iter()
+            .flat_map(|g| g.entries.iter())
+            .flat_map(|g| g.entries())
+            .collect::<Vec<_>>();
+
+        // For each module, the indices in the bitmap store which merge group entry modules
+        // transitively import that module. The bitmap can be treated as an opaque value, merging
+        // all modules with the same bitmap.
+        let mut module_merged_groups: FxHashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper> =
+            FxHashMap::with_capacity_and_hasher(module_count, Default::default());
+        // Entries that started a new merge group for some deopt reason
+        let mut entry_modules =
+            FxHashSet::with_capacity_and_hasher(module_count, Default::default());
+
+        let mergeable = graphs
+            .iter()
+            .flat_map(|g| g.iter_nodes())
+            .map(async |n| {
+                let module = n.module;
+                let mergeable = ResolvedVc::try_downcast::<Box<dyn MergeableModule>>(module);
+                if let Some(mergeable) = mergeable
+                    && *mergeable.is_mergeable().await?
+                {
+                    return Ok(Some(module));
+                }
+                Ok(None)
+            })
+            .try_flat_join()
+            .await?
+            .into_iter()
+            .collect::<FxHashSet<_>>();
+
+        let mut next_index = 0u32;
+        let visit_count = module_graph
+            .traverse_edges_fixed_point_with_priority(
+                entries.iter().map(|e| (*e, 0)),
+                &mut (),
+                |parent_info: Option<(&'_ SingleModuleGraphModuleNode, &'_ RefData)>,
+                 node: &'_ SingleModuleGraphModuleNode,
+                 _|
+                 -> Result<GraphTraversalAction> {
+                    // On the down traversal, establish which edges are mergable and set the list
+                    // indices.
+                    let (parent_module, hoisted) =
+                        parent_info.map_or((None, false), |(node, ty)| {
+                            (
+                                Some(node.module),
+                                match &ty.chunking_type {
+                                    ChunkingType::Parallel { hoisted, .. } => *hoisted,
+                                    _ => false,
+                                },
+                            )
+                        });
+                    let module = node.module;
+
+                    Ok(if parent_module.is_some_and(|p| p == module) {
+                        // A self-reference
```

Comment:
huh, i honestly didn't know that esm-self imports were allowed.  is that what this case is about, or is there something more subtle going on?